File: Tableau-Supported/Python/delete_data_in_existing_hyper_file.py
# -----------------------------------------------------------------------------
#
# This file is the copyrighted property of Tableau Software and is protected
# by registered patents and other applicable U.S. and international laws and
# regulations.
#
# You may adapt this file and modify it to fit into your context and use it
# as a template to start your own projects.
#
# -----------------------------------------------------------------------------
import shutil

from pathlib import Path

from tableauhyperapi import HyperProcess, Telemetry, \
    Connection, CreateMode, \
    NOT_NULLABLE, NULLABLE, SqlType, TableDefinition, \
    Inserter, \
    escape_name, escape_string_literal, \
    HyperException


def run_delete_data_in_existing_hyper_file():
    """
    An example of how to delete data in an existing Hyper file.
    """
    print("EXAMPLE - Delete data from an existing Hyper file")

    # Path to a Hyper file containing all data inserted into Customer, Product, Orders and LineItems table.
    # See "insert_data_into_multiple_tables.py" for an example that works with the complete schema.
    path_to_source_database = Path(__file__).parent / "data" / "superstore_sample.hyper"

    # Make a copy of the superstore example Hyper file.
    path_to_database = Path(shutil.copy(path_to_source_database, "superstore_sample_delete.hyper")).resolve()

    # Starts the Hyper Process with telemetry enabled to send data to Tableau.
    # To opt out, simply set telemetry=Telemetry.DO_NOT_SEND_USAGE_DATA_TO_TABLEAU.
    with HyperProcess(telemetry=Telemetry.SEND_USAGE_DATA_TO_TABLEAU) as hyper:

        # Connect to existing Hyper file "superstore_sample_delete.hyper".
        with Connection(endpoint=hyper.endpoint,
                        database=path_to_database) as connection:

            print(f"Delete all rows from customer with the name 'Dennis Kane' from table {escape_name('Orders')}.")
            # `execute_command` executes a SQL statement and returns the impacted row count.
            row_count = connection.execute_command(
                command=f"DELETE FROM {escape_name('Orders')} "
                f"WHERE {escape_name('Customer ID')} = ANY("
                f"SELECT {escape_name('Customer ID')} FROM {escape_name('Customer')} "
                f"WHERE {escape_name('Customer Name')} = {escape_string_literal('Dennis Kane')})")

            print(f"The number of deleted rows in table {escape_name('Orders')} "
                  f"is {row_count}.\n")

            print(f"Delete all rows from customer with the name 'Dennis Kane' from table {escape_name('Customer')}.")
            row_count = connection.execute_command(
                command=f"DELETE FROM {escape_name('Customer')} "
                f"WHERE {escape_name('Customer Name')} = {escape_string_literal('Dennis Kane')}")

            print(f"The number of deleted rows in table Customer is {row_count}.")

        print("The connection to the Hyper file has been closed.")
    print("The Hyper process has been shut down.")


if __name__ == '__main__':
    try:
        run_delete_data_in_existing_hyper_file()
    except HyperException as ex:
        print(ex)
        exit(1)

File: Tableau-Supported/CPP/delete_data_in_existing_hyper_file.cpp
/**
 * \example delete_data_in_existing_hyper_file.cpp
 *
 * An example of how to delete data in an existing Hyper file.
 */

#include <fstream>
#include <hyperapi/hyperapi.hpp>
#include <iostream>
#include <string>

/**
 * Helper function to copy a file
 */

static void copy(const std::string& sourcePath, const std::string& destinationPath) {
   std::ifstream source(sourcePath, std::ios::binary);
   std::ofstream destination(destinationPath, std::ios::binary);
   destination << source.rdbuf();
   source.close();
   destination.close();
}

static void runDeleteDataInExistingHyperFile() {
   std::cout << "EXAMPLE - Delete data from an existing Hyper file" << std::endl;

   // Path to a Hyper file containing all data inserted into Customer, Product, Orders and LineItems table
   // See "insert_data_into_multiple_tables.cpp" for an example that works with the complete schema.
   const std::string pathToSourceDatabase = "data/superstore_sample.hyper";

   // Make a copy of the superstore example Hyper file.
   const std::string pathToDatabase = "data/superstore_sample_delete.hyper";
   copy(pathToSourceDatabase, pathToDatabase);

   // Starts the Hyper Process with telemetry enabled to send data to Tableau.
   // To opt out, simply set telemetry=hyperapi::Telemetry::DoNotSendUsageDataToTableau.
   {
      hyperapi::HyperProcess hyper(hyperapi::Telemetry::SendUsageDataToTableau);

      // Connect to existing Hyper file "superstore_sample_delete.hyper".
      {
         hyperapi::Connection connection(hyper.getEndpoint(), pathToDatabase);

         std::cout << "Delete all rows from customer with the name 'Dennis Kane' from table " << hyperapi::escapeName("Orders") << "." << std::endl;
         // `executeCommand` executes a SQL statement and returns the impacted row count.
         int64_t rowCount = connection.executeCommand(
            "DELETE FROM " + hyperapi::escapeName("Orders") + " WHERE " + hyperapi::escapeName("Customer ID") + " = ANY(SELECT " +
            hyperapi::escapeName("Customer ID") + " FROM " + hyperapi::escapeName("Customer") + " WHERE " + hyperapi::escapeName("Customer Name") + " = " +
            hyperapi::escapeStringLiteral("Dennis Kane") + ")");
         std::cout << "The number of deleted rows in table " << hyperapi::escapeName("Orders") << " is " << rowCount << "." << std::endl
                   << std::endl;

         std::cout << "Delete all rows from customer with the name 'Dennis Kane' from table " << hyperapi::escapeName("Customer") << "." << std::endl;
         rowCount = connection.executeCommand(
            "DELETE FROM " + hyperapi::escapeName("Customer") + " WHERE " + hyperapi::escapeName("Customer Name") + " =" +
            hyperapi::escapeStringLiteral("Dennis Kane"));

         std::cout << "The number of deleted rows in table Customer is " << rowCount << "." << std::endl;
      }
      std::cout << "The connection to the Hyper file has been closed." << std::endl;
   }
   std::cout << "The Hyper Process has been shut down." << std::endl;
}

int main() {
   try {
      runDeleteDataInExistingHyperFile();
   } catch (const hyperapi::HyperException& e) {
      std::cout << e.toString() << std::endl;
      return 1;
   }
   return 0;
}
